design/one-pager-provider-nop..md

```diff
@@ -0,0 +1,91 @@
+- test creates an XRD and a composition of 3 nop objects
+  - each of those nop objects is told to become ready after 3 seconds
+- test creates an XR instance of this XRD
+- test waits and verifies that 3 nops objects are created, they all make it to the ready status, and the XR itself then makes it to the ready status because all 3 of its composed resources are now ready.
+
+This provider does exactly what it is told to. It can be used for testing functionality in core Crossplane without needing to account for the requirements of external APIs. It supports a single managed resource that contains a spec that allows for specifying how and when the status of the resource is changed. It does not orchestrate any external system. 
+
+This provider will be installed while running Crossplane end-to-end tests to satisfy any kind of composite infrastructure resource to simulate different scenarios. 
+
+## Design
+
+The provider will have a single managed resource `NopResource` that will reflect whatever is in the spec into the status. It will allow to define the resource to:
+- be ready after this period
+- be unhealthy after this period
+
+The idea is to have an array of fields which will let you declare condition type and status of the resource at each time interval. 
+
+This will be implemented by making use of three fields:
+- `conditionType` and `conditionStatus` for declaring condtion of resource
+- `timeAfter` for declaring the time elapsed after the creation of resource at which we need to set the specified condition.
+
+The resource structure might look something like: 
+```go
+type ResourceConditionAfter struct {
+  Time            string `json:"time"`
+  ConditionType   string `json:"conditionType"`
+  ConditionStatus string `json:"conditionStatus"`
+}
+
+type NopResourceParameters struct {
+	ConditionAfter []ResourceConditionAfter `json:"conditionAfter"`
+}
+```
+
+### `NopResource` Controller
+
+The controller for this nop type will not call any external APIs during its reconcile loop and just perform the behaviour as specified by the user in the spec. 
+
+#### Setup
+Since default is too long for the testing use case, the sync period should be decreased to a small value like `100ms` using `WithPollInterval`in the Setup function. This will allow controller to set various conditions.
+
+#### Connect
+Currently, there is no requirement for having `ProviderConfig` since we don't have any external API client that we might want to connect. 
+We would just create an external client with no service parameter passed. 
+
+#### Observe
+This is where the main logic will reside that would compare the time elapsed with the time interval passed in the spec. As the time elapsed becomes equal to the time passed it will change the condition of resource accordingly. 
+
+#### Update
+Currently there is no plan to update the NopResource so this function will just set condition to crossplane-runtime `Available`. 
```

Comment:
This doesn't sound correct to me, because it would likely be in conflict with the status conditions we are supposed to be setting according to the desired behavior from the spec.  e.g. if spec says ready: false after 5 seconds, we wouldn't want the Update function to set it to Available, that would effectively wipe out the ready: false that we desire.

One potential flow that may be aligned with the design of the managed reconciler would be:

1) Observe function should determine if the status conditions match what the spec says they should be.  If they do not match (i.e. we need to update the status conditions), it can return `ResourceUpToDate: false`, indicating that Update should be called to perform the update.  See more about this value in https://github.com/crossplane/provider-template/blob/master/internal/controller/mytype/mytype.go#L145-L148
2) Update function should then carry out the intended status/conditions change as per what the spec is declaring.  It makes more sense in my mind that updating the resource/status should be done in the Update function.

Thoughts?

Reply:
Yes, that makes perfect sense to me @jbw976. I also was having second thoughts about this and hence thought to jot it down so that we could discuss it. Let's go with the above-mentioned flow 👍🏼